### PR TITLE
optimize: avoid a duplicate string length lookup

### PR DESCRIPTION
**Pull request by p0pr0ck5**

### CODE REVIEW

File: lib/resty/string.lua [hunk 1]
```diff
@@ -27,7 +27,7 @@ local str_type = ffi.typeof("uint8_t[?]")
 function _M.to_hex(s)
     local len = #s * 2
     local buf = ffi_new(str_type, len)
-    C.ngx_hex_dump(buf, s, #s)
+    C.ngx_hex_dump(buf, s, len / 2)
```
[the remark this answers is not in the document]
Comment (p0pr0ck5) [reply lib/resty/string.lua]: makes sense. done.

[the remark this answers is not in the document]
Comment (p0pr0ck5) [reply lib/resty/string.lua]: well then you have to either add another local env or still use two mul/div calcs:

```
    local len = #s
    local buf = ffi_new(str_type, len * 2)
    C.ngx_hex_dump(buf, s, len)
    return ffi_str(buf, len * 2)
```

or

```
    local len = #s
    local blen = s * 2
    local buf = ffi_new(str_type, blen)
    C.ngx_hex_dump(buf, s, len)
    return ffi_str(buf, blen)
```


thoughts?

[the remark this answers is not in the document]
Comment (dndx) [reply lib/resty/string.lua]: Let's do the second form. I think it would be the fastest.

[the remark this answers is not in the document]
Comment (p0pr0ck5) [reply lib/resty/string.lua]: makes sense. done.

File: lib/resty/string.lua [hunk 2]
```diff
@@ -25,10 +25,11 @@ local str_type = ffi.typeof("uint8_t[?]")
 
 
 function _M.to_hex(s)
-    local len = #s * 2
-    local buf = ffi_new(str_type, len)
-    C.ngx_hex_dump(buf, s, #s)
-    return ffi_str(buf, len)
+    local len = #s
+    local buf_len = s * 2
```
Comment (dndx) [inline lib/resty/string.lua]: This is causing Lua errors in the Travis tests. Should be `len * 2`.

Comment (p0pr0ck5) [reply lib/resty/string.lua]: just caught this on my end as well. doh :| sorry. fixed.

**Comment by thibaultcha** [comment]
Nice one, thanks!

**Comment by thibaultcha** [comment]
Oh, and thank you @dndx for the review ;)

**Pull request merged**